modules/tinymce/src/core/main/ts/newline/InsertNewLine.ts

```diff
@@ -9,23 +9,31 @@ import { Fun, Type } from '@ephox/katamari';
 
 import Editor from '../api/Editor';
 import { EditorEvent } from '../api/util/EventDispatcher';
-import { fireFakeInputEvent } from '../keyboard/FakeInputEvents';
+import { fireFakeBeforeInputEvent, fireFakeInputEvent } from '../keyboard/FakeInputEvents';
 import * as InsertBlock from './InsertBlock';
 import * as InsertBr from './InsertBr';
 import * as NewLineAction from './NewLineAction';
 
 const insert = (editor: Editor, evt?: EditorEvent<KeyboardEvent>) => {
   NewLineAction.getAction(editor, evt).fold(
     () => {
-      InsertBr.insert(editor, evt);
-      if (Type.isNonNullable(evt)) {
-        fireFakeInputEvent(editor, 'insertLineBreak');
+      const event = fireFakeBeforeInputEvent(editor, 'insertLineBreak');
```

Comment:
This logic doesn't look correct... it should not be firing the `beforeinput` event here if `evt` is nullable, as `input` also won't fire if `evt` is nullable.

- [ ] Fix this and below to only fire `beforeInput` if `evt` is nonnullable

Reply:
However, we change the DOM on the `InsertBr.Insert`, don't we? Which we don't want to do before the beforeinput, or if it's cancelled.

Reply:
I'm not sure what you mean here... it's not related as to whether or not we change the DOM. It's related to if this was triggered by an event, so if this is triggered programattically then we shouldn't be firing `beforeinput` or `input` events.

Reply:
No, I mean, I assumed we wanted to be able to cancel the input and prevent changes to the DOM, whereas this one would be un-blockable even when you get the event.
I'm not saying that the event shouldn't be moved not to fire when not appropriate, just that we shouldn't change the DOM even if the input is cancelled? Or did I get something wrong on that?

Reply:
You can't cancel it using the event in that case, because it's not been triggered by an event. As I said, you should only be able to do that when this logic is triggered via an event.

Reply:
As an example, this logic can be triggered via the `mceInsertNewLine` so firing an event there makes no sense and instead if we wanted that cancelled you'd cancel the command instead.